crates/red_knot_python_semantic/src/types/mro.rs

```diff
@@ -0,0 +1,641 @@
+            //
+            // For a Python-3 translation of the algorithm described in that document,
+            // see https://gist.github.com/AlexWaygood/674db1fce6856a90f251f63e73853639
+            _ => {
+                let possible_mros_per_base: Vec<(ClassBase, Cow<MroPossibilities>)> = bases_possibility
+                    .iter()
+                    .map(|base| (*base, base.mro_possibilities(db)))
+                    .collect();
+
+                let mro_cartesian_product = possible_mros_per_base
+                    .iter()
+                    .map(|(base, mro_possibilities)| mro_possibilities.iter(db, *base))
+                    .multi_cartesian_product();
+
+                // Each `possible_mros_of_bases` is a concrete possibility of the list of mros of all of the bases:
+                // where the bases are `[B1, B2, B..N]`, `possible_mros_of_bases` represents one possibility of
+                // `[mro_of_B1, mro_of_B2, mro_of_B..N]`
+                for possible_mros_of_bases in mro_cartesian_product {
+                    let possible_mros_of_bases: Vec<VecDeque<ClassBase>> = possible_mros_of_bases
+                        .into_iter()
+                        .map(|mro|mro.iter().copied().collect())
+                        .collect();
+                    let linearized = c3_merge(
+                        std::iter::once(VecDeque::from([ClassBase::Class(class)]))
+                            .chain(possible_mros_of_bases)
+                            .chain(std::iter::once(bases_possibility.iter().copied().collect()))
+                            .collect(),
+                    );
+                    match linearized {
+                        Some(mro) => {
+                            mro_possibilities.insert(mro);
+                        },
+                        None => {
+                            if !mro_errors.contains(&bases_possibility) {
+                                mro_errors.insert(bases_possibility.clone());
+                            }
+                        },
+                    };
+                }
+            }
+        }
+    }
+
+    MroPossibilities::possibly_many(mro_possibilities, mro_errors)
+}
+
+/// Given a list of types representing the bases of a class,
+/// of which one or more types could be a [`Type::Union`] variant,
+/// resolve the list into a "union of bases lists", where each list in the union
+/// is guaranteed not to hold any bases that are a [`Type::Union`].
+fn fork_bases<'db>(db: &'db dyn Db, bases: &[Type<'db>]) -> BasesPossibilities<'db> {
+    // Fast path for the common case, where none of the bases is a [`Type::Union`]:
+    if !bases.iter().any(Type::is_union) {
+        return BasesPossibilities::Single(bases.iter().map(ClassBase::from).collect());
+    }
+
+    // Slow path: one or more of the bases is a [`Type::Union`]
+    let mut possibilities = FxHashSet::from_iter([BasesList::default()]);
```

Comment:
If a class does not explicitly inherit from anything, Python synthesizes a `__bases__` tuple consisting solely of `object`. This is because all classes in Python are subclasses of `object`, which is the top type, the supertype of everything else.

```pycon
>>> class Foo: pass
... 
>>> Foo.__bases__
(<class 'object'>,)
```